scheduler/src/cook/scheduler/constraints.clj

```diff
@@ -99,33 +90,48 @@
+       first))
+
+
+(defrecord gpu-host-constraint [job]
   JobConstraint
   (job-constraint-name [this] (get-class-name this))
   (job-constraint-evaluate
     [this _ vm-attributes]
     (job-constraint-evaluate this nil vm-attributes []))
   (job-constraint-evaluate
-    [this _ vm-attributes target-vm-tasks-assigned]
-    (let [; Look at attribute and running jobs to determine if vm has gpus
-          vm-has-gpus? (or (get vm-attributes "COOK_GPU?") ; Set when putting attributes in cache
-                           (some (fn gpu-task? [{:keys [needs-gpus?]}]
-                                   needs-gpus?)
-                                 target-vm-tasks-assigned))
-          job (:job this)
-          passes? (or (and needs-gpus? vm-has-gpus?)
-                      (and (not needs-gpus?) (not vm-has-gpus?)))]
-      [passes? (when-not passes? (if (and needs-gpus? (not vm-has-gpus?))
-                                   "Job needs gpus, host does not have gpus."
-                                   "Job does not need gpus, host has gpus."))])))
+    [{:keys [job]} _ vm-attributes target-vm-tasks-assigned]
+    (let [gpu-model-requested (or (-> job util/job-ent->env (get "COOK_GPU_MODEL"))
+                                  (:default-model (gpu-models-config-for-pool (config/valid-gpu-models) (util/job->pool-name job))))
+          gpu-count-requested (-> job util/job-ent->resources :gpus (or 0))
+          gpu-model->count-available (get vm-attributes "gpus")
+          is-k8s-vm (= (get vm-attributes "compute-cluster-type") "kubernetes")
+          ; k8s VM that supports gpu models but does not have any available gpus will have a gpus map of {"model A" 0 "model B" 0 ...}
+          ; k8s VM that does not support gpu models will have an empty gpus map of {}
+
+          ; requesting gpus is only being supported for kubernetes-bound jobs
+          passes? (if is-k8s-vm
+                    (if (and (pos? gpu-count-requested) gpu-model-requested)
+                      (>= (get gpu-model->count-available gpu-model-requested 0) gpu-count-requested)
+                      ; if job is kubernetes-bound and does not request gpus, do not schedule it on a VM that supports gpus
+                      (-> gpu-model->count-available count zero?))
+                    ; if job is mesos-bound, ensure that the gpu-count-requested is 0
+                    (zero? gpu-count-requested))]
+      [passes? (when-not passes? (if (not gpu-model-requested)
+                                   "Job does not need GPUs, host has GPUs."
+                                   "Job needs GPUs that are not present on host"))])))
 
 (defn build-gpu-host-constraint
   "Constructs a gpu-host-constraint.
-  The constraint prevents a gpu job from running on a non-gpu host (resources should also handle this)
+  The constraint prevents a gpu job from running on a host that does not have the correct number and model of gpus (resources should also handle this)
```

Comment:
is the `(resources should also handle this)` note still relevant?

Reply:
I don't think so

Reply:
ok, deleted